bayes_kit/ess.py

```diff
@@ -1,66 +1,26 @@
 import numpy as np
 import numpy.typing as npt
+import bayes_kit.autocorr as autocorr
 
 FloatType = np.float64
 IntType = np.int64
 VectorType = npt.NDArray[FloatType]
 
-def autocorr_fft(chain: VectorType) -> VectorType:
-    """
-    Return sample autocorrelations at all lags for the specified sequence.
-    Algorithmically, this function calls a fast Fourier transform (FFT).
-
-    Parameters:
-    chain: sequence whose autocorrelation is returned
-
-    Returns:
-    autocorrelation estimates at all lags for the specified sequence
-    """
-    size = 2 ** np.ceil(np.log2(2 * len(chain) - 1)).astype("int")
-    var = np.var(chain)
-    ndata = chain - np.mean(chain)
-    fft = np.fft.fft(ndata, size)
-    pwr = np.abs(fft) ** 2
-    N = len(ndata)
-    acorr = np.fft.ifft(pwr).real / var / N
-    return acorr
-
-def autocorr_np(chain: VectorType) -> VectorType:
-    """
-    Return sample autocorrelations at all lags for the specified sequence.
-    Algorithmically, this function delegates to the Numpy `correlation()` function.
-
-    Parameters:
-    chain: sequence whose autocorrelation is returned
-
-    Returns:
-    autocorrelation estimates at all lags for the specified sequence
-    """
-    chain_ctr = chain - np.mean(chain)
-    N = len(chain_ctr)
-    acorrN = np.correlate(chain_ctr, chain_ctr, "full")[N - 1 :]
-    return acorrN / N
-
-def autocorr(chain: VectorType) -> VectorType:
-    """
-    Return sample autocorrelations at all lags for the specified sequence.
-    Algorithmically, this function delegates to `autocorr_fft`.
-
-    Parameters:
-    chain: sequence whose autocorrelation is returned
-
-    Returns:
-    autocorrelation estimates at all lags for the specified sequence
-    """
-    return autocorr_fft(chain)
-
-def first_neg_pair_start(chain: VectorType) -> IntType:
-    """
-    Return the index of first element of the sequence whose sum with the following
-    element is negative, or the length of the sequence if there is no such element.
+def _end_pos_pairs(chain: VectorType) -> IntType:
+    """Return the index of first element of the sequence whose sum with
```

Comment:
"Return the index of the first *even-indexed* element" maybe? It's implicit in the example and the comment about "indexes that come in positive pairs" but as written this sentence doesn't mention index parity, so a quick skim would give the wrong idea.